kernel/sched.c

```diff
@@ -691,6 +691,42 @@ int z_pend_curr(struct k_spinlock *lock, k_spinlock_key_t key,
 	 * held.
 	 */
 	(void) k_spin_lock(&_sched_spinlock);
+
+#if defined(CONFIG_KERNEL_WARN_LONG_TIME_PENDING)
+	int seconds = (_current)->long_time_warns;
+
+	if (seconds >= 0 &&
+	    (K_TIMEOUT_EQ(timeout, K_FOREVER) ||
+	     (timeout.ticks > K_SECONDS(seconds).ticks))) {
+		uint32_t ref = k_uptime_get_32(), end;
+		k_ticks_t elipsed;
+		int ret;
+
+		pend_locked(_current, wait_q, K_SECONDS(seconds));
+		k_spin_release(lock);
+
+		ret = z_swap(&_sched_spinlock, key);
+		if (ret != -EAGAIN) {
+			return ret;
+		}
+
+		end = k_uptime_get_32();
+
+		LOG_WRN("%s blocked for %u secs",
+			k_thread_name_get(_current), (end - ref) / MSEC_PER_SEC);
+
```

Comment:
This retry logic is fraught with peril, and gets a hard -1 from me.

At the root of it, there is no guarantee that when the thread in question becomes ready to run again is when it will actually be scheduled to execute. This creates a gap where a number things can go seriously wrong.

1. The resource on which it was pending may become available and then given to a different thread that was waiting. This will violate the expected order in which threads obtain resources.
2. The resource on which it was pending may become available and if there was no one in the wait queue, the resource's internals may get updated before the thread re-pends. This can lead to corrupted states. (Imagine a semaphore whose count has increased because it does not have a waiter, but then the thread pends leading to a condition where we have a thread waiting on an available semaphore).
3. As the duration of this gap is unknown, this has an implicit impact to what developers understand as a timeout. Setting aside the extra tick that the system adds to account for partial ticks, it is something like the maximum length of contiguous time that a thread will compete for a resource before giving up on obtaining it. (It can probably be worded better.)